### ISSUE TITLE
How do base_weights work in inference time?

### ISSUE THREAD

**Issue by lucidfrontier45** [dmlc/xgboost#11521]
I'm trying to create my own minimal inference engine without direct dependency of XGBoost itself. My final goal is to implement it in Rust but here I first test the concept in Python.

I started from understanding the exported JSON format.
My assumption is that `Booster.predict` returns an output value for a sample by summing all corresponding leaf values (which is `base_weights`) of all trees and plus base score.
So I first tried to check what are all possible output values.

1. For a Booster trained as a regressor, my scripts returns values that contains exactly the same values as Booster.predict.
2. However for a classifier booster, my scripts and Booster.predict's result didn't match
3. Supprisingly, I expected if I set all `base_weights` to zero, all output values will be exactly the same as `base_score` by Booster.predict returned many different values.

Could you explain which part of my assumption was wrong and what is the correct way to do inference?

I put train, eval scripts as well as three model JSON files to this gist.
https://gist.github.com/lucidfrontier45/3948d8a912ece8dcaada2dd88d3329ba

The XGBoost version is 3.0.2.

Thank you.

[Edit 2025/06/24]

I figured out that for binary:logistic objective, Booster.predict returns

```
sigmoid( logit(base_score) + sum(leaf_values))
```

and I successfully reproduce the same result as XGBoost's own inference result.

I also found that `split_conditions` and `base_weights` values are all the same in the JSON file and surprisingly  `split_conditions` values are actually used for leaf values.  But then what is the purpose of `base_weights`?

**Comment by trivialfis** [comment]
A high-level description can be found here https://xgboost.readthedocs.io/en/stable/tutorials/intercept.html . Since you have figured out the binary logistic, I think rest should be easy if you plug your equations of binary logistic into the document. But do let me know if you run into issues of doing so, I can help improve the document.

> split_conditions values are actually used for leaf values.

Yes, it's an implementation detail. We reuse the same buffer for split value and leaf value. You can check whether a node is a leaf.

>  But then what is the purpose of base_weights?

Depends on parameters like learning rate.

**Comment by lucidfrontier45** [comment]
@trivialfis 

Thank you for kind response.

Regarding base_weights and learning rate, I don't see there is learning rate related parameters in the model's JSON Schema.
https://xgboost.readthedocs.io/en/stable/tutorials/saving_model.html#json-schema

Does it mean I can always ignore `base_weights` and just use `split_conditions` if I want to load model JSON and develop an inference engine?

**Comment by trivialfis** [comment]
Yes, you can always ignore the base weight. The base weight is before the application of the learning rate (leaf_weight = base_weight * learning_rate).
